pkg/controller/batchrelease/workloads/cloneset_control_plane.go

```diff
@@ -138,26 +139,29 @@ func (c *CloneSetRolloutController) UpgradeOneBatch() (bool, error) {
 		int(c.releaseStatus.ObservedWorkloadReplicas), true)
 
 	// in case of no need to upgrade pods
-	if currentCanaryReplicas >= canaryGoal && int32(workloadPartition) <= stableGoal {
-		klog.V(3).InfoS("upgraded one batch, but no need to update partition of cloneset",
-			"BatchRelease", c.releasePlanKey, "current-batch", c.releaseStatus.CanaryStatus.CurrentBatch,
-			"canary-goal", canaryGoal, "stable-goal", stableGoal, "canary-replicas", currentCanaryReplicas, "partition", workloadPartition)
-		return true, nil
-	}
-
-	// if canaryReplicas is int, then we use int;
-	// if canaryReplicas is percentage, then we use percentage.
-	var partitionGoal intstr.IntOrString
-	canaryIntOrStr := c.releasePlan.Batches[c.releaseStatus.CanaryStatus.CurrentBatch].CanaryReplicas
-	if canaryIntOrStr.Type == intstr.Int {
-		partitionGoal = intstr.FromInt(int(stableGoal))
-	} else if c.releaseStatus.ObservedWorkloadReplicas > 0 {
-		partitionGoal = ParseIntegerAsPercentageIfPossible(stableGoal, c.releaseStatus.ObservedWorkloadReplicas, &canaryIntOrStr)
+	IsUpgradedDone := func() bool {
+		return currentCanaryReplicas >= canaryGoal && int32(workloadPartition) <= stableGoal
 	}
+	if !IsUpgradedDone() {
+		// if canaryReplicas is int, then we use int;
+		// if canaryReplicas is percentage, then we use percentage.
+		var partitionGoal intstr.IntOrString
+		canaryIntOrStr := c.releasePlan.Batches[c.releaseStatus.CanaryStatus.CurrentBatch].CanaryReplicas
+		if canaryIntOrStr.Type == intstr.Int {
+			partitionGoal = intstr.FromInt(int(stableGoal))
+		} else if c.releaseStatus.ObservedWorkloadReplicas > 0 {
+			partitionGoal = ParseIntegerAsPercentageIfPossible(stableGoal, c.releaseStatus.ObservedWorkloadReplicas, &canaryIntOrStr)
+		}
 
-	// upgrade pods
-	if err := c.patchCloneSetPartition(c.clone, &partitionGoal); err != nil {
+		// upgrade pods
+		if err := c.patchCloneSetPartition(c.clone, &partitionGoal); err != nil {
+			return false, err
+		}
 		return false, nil
```

Comment:
if we return directly in the "! IsUpgradedDone" condition, no message will be logged, then why bother reversing the if else clause?